`Development`: Add Helios-based deployment workflow for test servers and reusable build logic

egekocabas opened the pull request:
### Checklist

#### General

- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

### Motivation and Context
This PR adds a new action file `testserver-deployment.yml` to integrate with the Helios web application for test server deployments. The current **label-based approach** for triggering deployments will be replaced with this new `workflow_dispatch` based workflow.

Additionally, the `build.yml` workflow has been refactored to leverage a reusable workflow file (`reusable-build.yml`). This ensures consistency across workflows and simplifies future maintenance.

### Description

**Integration with Helios:**
- The workflow is designed to work with the Helios application, which manages deployment requests.
- Helios handles locking of test servers and tracking deployment statuses.
- Helios stays in sync with GitHub through webhook events and periodic automatic syncing, ensuring all repository changes, such as new pull requests or branch updates, are accurately reflected in the Helios interface.

**Reusable Build Workflow:**

- Introduced `reusable-build.yml` to streamline build logic.
- The `build.yml` workflow has been updated to utilize this reusable workflow.
- The new `testserver-deployment.yml` workflow also utilizes the reusable workflow to handle deployment scenarios where builds are not found.

**testserver-deployment.yml:**
- Replaced label-based deployment triggers with a `workflow-dispatch` approach.
- Enabled deployments for both PRs and standalone branches.
- The label-based approach and other related logic will be removed in another PR, which will consolidate the contents of `testserver-deployment.yml` into `testserver.yml`.


Once this PR is merged, a follow-up PR will finalize the transition by moving the contents of `testserver-deployment.yml` into `testserver.yml`.

egekocabas commented:
I will add a new GitHub action so that we can merge this PR before release of Helios. And with that Artemis be ready for Helios before removing all the label related functionality.

Hialus commented on .github/workflows/testserver-helios.yml:
```diff
@@ -0,0 +1,364 @@
+          exit 1
+      - name: Post comment about build validation failure
+        if: failure() && github.event.inputs.HELIOS_BUILD == 'false'
+        uses: marocchino/sticky-pull-request-comment@v2
```
Why are you using a third-party action for this? Posting comments is very easily done with plain Github script. Generally, you should only use third-party actions when they do things that are otherwise not easily doable. Especially in the deployment workflow that handles sensitive secrets, I'd want you to refrain from using third-party actions unless absolutely needed.

This applies to all usages of this action.

egekocabas replied:
This was used for **sticky** comments, where we only wanted to comment once and update/remove if comment exist with the header definition, this third-part action handles that. Right now i removed commenting logic and in later we can add and handle this with our own logic.

Hialus commented on .github/workflows/build.yml:
```diff
@@ -31,6 +31,9 @@ on:
     types:
     - created
 
+
+# Keep build.yml in sync with testserver-helios.yml
+
```
What has to be kept in sync? Please be more specific

Hialus commented on .github/workflows/testserver-helios.yml:
```diff
@@ -0,0 +1,364 @@
+        description: "Docker tag to use if we are building or pulling an existing image"
+        required: false
+        type: string
+      HELIOS_RAW_URL:
+        description: "URL to the raw content of the repository in the format https://raw.githubusercontent.com/:owner/:repo/:sha"
+        required: false
+        type: string
```
Why is this not marked as required? If this is not present the workflow would fail.

Hialus commented on .github/workflows/testserver-helios.yml:
```diff
@@ -0,0 +1,364 @@
+    runs-on: ubuntu-latest
+    steps:
+      # Build the Docker image (branch without PR)
+      - name: Git Checkout
+        if: ${{ github.event.inputs.HELIOS_BUILD == 'true' }}
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ github.event.inputs.HELIOS_BRANCH_HEAD_SHA }}
+      - name: Set up QEMU
+        if: ${{ github.event.inputs.HELIOS_BUILD == 'true' }}
+        uses: docker/setup-qemu-action@v3
+      - name: Set up Docker Buildx
+        if: ${{ github.event.inputs.HELIOS_BUILD == 'true' }}
+        uses: docker/setup-buildx-action@v3
+      # Build and Push to GitHub Container Registry
+      - name: Login to GitHub Container Registry
+        if: ${{ github.event.inputs.HELIOS_BUILD == 'true' }}
+        uses: docker/login-action@v3
+        with:
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+      - name: Build and Push to GitHub Container Registry
+        if: ${{ github.event.inputs.HELIOS_BUILD == 'true' }}
+        uses: docker/build-push-action@v5
+        with:
+          # beware that the linux/arm64 build from the registry is using an amd64 compiled .war file as
+          # the GitHub runners don't support arm64 and QEMU takes too long for emulating the build
+          platforms: linux/amd64,linux/arm64
+          file: ./docker/artemis/Dockerfile
+          context: .
+          tags: ghcr.io/ls1intum/artemis:${{ github.event.inputs.HELIOS_BUILD_TAG }}
+          push: true
+          cache-from: type=gha
+          cache-to: type=gha,mode=min
```
This could be a separate job. That way you would only need one `if: ${{ github.event.inputs.HELIOS_BUILD == 'true' }}` and it would be better structured.

Hialus replied:
Ideally you would also reuse the existing build workflow instead of copying it over. This could be done via reusable workflows.

Hialus replied:
Also this will result in deployments that take an extra 10 minutes of runtime. During this time no other deployment can run. This is pretty bad and you should find a solution to this.

egekocabas commented:
@Hialus Thank you for your review! I have made the necessary changes and wanted to provide some details about them:

- `workflow_call` in `build.yml`
  - Added `workflow_call` to `build.yml` to ensure it runs both for default events without breaking anything and can also be used as a reusable workflow.
  - The behavior of reusable workflows differs from the documentation in some key ways. For example, in our case, it’s not possible to check `{{ github.event_name == 'workflow_call' }}` because the parent workflow passes its event to the child (reusable) workflow. As a result, I can only check `{{ github.event_name == 'workflow_dispatch' }}`.
  - Additionally, when using `actions/github-script`, I can’t access the inputs of `workflow_call` directly but can access the parent’s inputs instead.
- `concurrency` for safe environment deployments:
  - Changed `concurrency` to `concurrency: ${{ github.event.inputs.HELIOS_ENVIRONMENT_NAME }}`. This ensures that deployments for different environments remain isolated, even if a conditional build takes over 10 minutes. Other environment deployments will not be affected since concurrency is now based on the environment.
- Handling dependencies for `deploy` job:
  - Since we either build (if no PR exists) or use an existing build (if a PR exists), the `deploy` job depends on two jobs, but one of them will always be skipped. Using the following setup directly would result in the deployment job being skipped if one dependency is skipped:
```
  deploy:
    needs: [ conditional-build, check-existing-build ]
```

    To handle this, I updated the configuration like this:

```
  deploy:
    needs: [ conditional-build, check-existing-build ]
    # Run if either job is successful
    # This if condition should be exactly like below, since one of the job is skipped
    if: always() && (needs.conditional-build.result == 'success' || needs.check-existing-build.result == 'success')
```
  - This ensures that the deploy job runs as long as one of the dependencies is successful.

Let me know if you have any further questions or suggestions!

egekocabas commented:
@Hialus 
Since the new GitHub Action file, `testserver-helios.yml`, hasn’t been pushed to the default branch yet, I tested the changes by temporarily replacing the content of the existing action file, `testserver.yml`, with the new workflow. After confirming that the workflow runs without any issues, I reverted the changes in `testserver.yml` and copied the updated content to the new file, `testserver-helios.yml`.

Below are two example runs of the deploy script:

With build: https://github.com/ls1intum/Artemis/actions/runs/12837428161
Without build: https://github.com/ls1intum/Artemis/actions/runs/12837536986

ole-ve commented on .github/workflows/testserver-helios.yml:
```diff
@@ -0,0 +1,161 @@
+        required: true
+        type: string
+      HELIOS_BRANCH_HEAD_SHA:
+        description: "SHA of the branch to deploy"
```
`
```suggestion
        description: "SHA (Head commit) of the branch to deploy"
```

ole-ve commented on .github/workflows/build.yml:
```diff
@@ -30,6 +30,25 @@ on:
+        required: true
+        type: string
+      raw_url:
+        description: "URL to use for fetching files in the format https://raw.githubusercontent.com/:owner/:repo/:sha"
```
it's a little bit unclear to me what `fetching files` means. AFAIS, we only use the `RAW_URL` for downloading the artemis-cli, right? would that work with this example URL?

egekocabas replied:
Thank you for the feedback! You are correct that the `RAW_URL` input is currently used in the [testserver.yml](https://github.com/ls1intum/Artemis/blob/develop/.github/workflows/testserver.yml#L296) workflow specifically for fetching the `artemis-server-cli`. However, I am not entirely certain whether the `RAW_URL` environment variable defined in the [build.yml](https://github.com/ls1intum/Artemis/blob/develop/.github/workflows/build.yml#L39) workflow is actively used during the build process.

The environment variable was already present before the changes, and since I wasn't familiar with its original purpose, I added the override in the `workflow_call` step to ensure consistency across workflows, just in case it is being used somewhere.

I agree the description for `raw_url` might be misleading.

If you have insights about whether `RAW_URL` is actively used in the build process (beyond setting it as an environment variable), I would appreciate it!

ole-ve reviewed (commented):
reviewed code, added two questions

krusche commented on .github/workflows/testserver-helios.yml:
```diff
@@ -0,0 +1,161 @@
+on:
+  workflow_dispatch:
+    inputs:
+      HELIOS_TRIGGERED_BY:
```
does it make sense to introduce HELIOS here? looks like a strong coupling, but ideally Artemis stays independent of the actual CI provider, so I would prefer HELIOS independent variables and names

Hialus commented on .github/workflows/build.yml:
```diff
@@ -44,7 +63,20 @@ jobs:
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v4
+      # If this is a workflow_dispatch, override RAW_URL
+    - name: Override RAW_URL for workflow_dispatch
+      if: ${{ github.event_name == 'workflow_dispatch' }}
+      run: echo "RAW_URL=${{ inputs.raw_url }}" >> $GITHUB_ENV
+      # If this workflow is called via `workflow_dispatch`, checkout the specified branch
+    - name: Git Checkout (for workflow_dispatch)
+      if: ${{ github.event_name == 'workflow_dispatch' }}
+      uses: actions/checkout@v4
+      with:
+        ref: ${{ inputs.sha }}
+      # Otherwise, checkout the default branch
+    - name: Git Checkout
+      if: ${{ github.event_name != 'workflow_dispatch' }}
+      uses: actions/checkout@v4
```
To avoid hacks like this, it would be better to:
1. Create one reusable workflow for building Artemis
2. Create a new workflow that has the triggers from the old build workflow and then call the reusable workflow with the proper inputs
3. Call the reusable workflow from your deployment workflow with the proper inputs

The current approach needlessly complicates the current build workflow. By doing a proper reusable workflow, it would stay about the same complexity.

Hialus commented on .github/workflows/build.yml:
```diff
@@ -81,12 +113,22 @@ jobs:
       with:
         result-encoding: string
         script: |
+          if (context.eventName === "workflow_dispatch") {
+            // When the workflow_call is triggered by another workflow,
+            // the parent's payload is passed to the child workflow
+            // This is why we are using the inputs from the parent workflow
+            return context.payload.inputs.HELIOS_BUILD_TAG;
+          }
```
I do not like the strong coupling to both Helios and the parent workflow here. This should just be passed via an input to the reusable workflow.

Hialus reviewed (changes requested):
Thanks for the changes. The testserver-helios workflow now looks okay. However, I do not like the implementation of the reusable workflow.

ole-ve reviewed (approved):
thanks for integrating the feedback. code looks good, really like the refactoring 👍
